x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java

```diff
@@ -40,19 +46,132 @@ final class ComputeListener implements Releasable {
+        ActionListener<ComputeResponse> delegate
+    ) {
         this.transportService = transportService;
         this.task = task;
         this.responseHeaders = new ResponseHeadersCollector(transportService.getThreadPool().getThreadContext());
         this.collectedProfiles = Collections.synchronizedList(new ArrayList<>());
+        this.esqlExecutionInfo = executionInfo;
+        this.queryStartTimeNanos = queryStartTimeNanos;
+        this.whereRunning = clusterAlias;
+        // for the DataNodeHandler ComputeListener, clusterAlias and executionInfo will be null
+        // for the top level ComputeListener in ComputeService both will be non-null
+        assert (clusterAlias == null && executionInfo == null) || (clusterAlias != null && executionInfo != null)
+            : "clusterAlias and executionInfo must both be null or both non-null";
+
+        // listener that executes after all the sub-listeners refs (created via acquireCompute) have completed
         this.refs = new RefCountingListener(1, ActionListener.wrap(ignored -> {
             responseHeaders.finish();
-            var result = new ComputeResponse(collectedProfiles.isEmpty() ? List.of() : collectedProfiles.stream().toList());
+            ComputeResponse result;
+
+            if (runningOnRemoteCluster()) {
+                // for remote executions - this ComputeResponse is created on the remote cluster/node and will be serialized and
+                // received by the acquireCompute method callback on the coordinating cluster
+                EsqlExecutionInfo.Cluster cluster = esqlExecutionInfo.getCluster(clusterAlias);
+                result = new ComputeResponse(
+                    collectedProfiles.isEmpty() ? List.of() : collectedProfiles.stream().toList(),
+                    cluster.getTook(),
+                    cluster.getTotalShards(),
+                    cluster.getSuccessfulShards(),
+                    cluster.getSkippedShards(),
+                    cluster.getFailedShards()
+                );
+            } else {
+                result = new ComputeResponse(collectedProfiles.isEmpty() ? List.of() : collectedProfiles.stream().toList());
+                if (coordinatingClusterIsSearchedInCCS()) {
+                    // mark local cluster as finished once the coordinator and all data nodes have finished processing
+                    executionInfo.swapCluster(
+                        RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY,
+                        (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL).build()
+                    );
+                }
+            }
             delegate.onResponse(result);
         }, e -> delegate.onFailure(failureCollector.getFailure())));
     }
 
+    /**
+     * @return true if the "local" querying/coordinator cluster is being searched in a cross-cluster search
+     */
+    private boolean coordinatingClusterIsSearchedInCCS() {
```

Comment:
I think we can remove the methods `coordinatingClusterIsSearchedInCCS`, `runningOnRemoteCluster`, and `isCCSListener`, `shouldRecordTookTime` as we should be able to collect execution info similarly to how we handle profiles and warnings. These methods require careful consideration of when and how to pass the `clusterAlias`, which I think increases the risk of errors. That said, it's entirely up to you whether to keep them as they are or remove them.

Reply:
I'm not sure that's true. I find the ComputeListener very hard to reason about, so I could be wrong. The core problems with the profiles/warnings approach is that they accumulate with an intent to be shown in the final response, but the EsqlExecutionInfo needs to be updated in real time as the search is running so that when we add `_clusters` to the async-search response of still running searches (in a later PR) it will have all the current data in the EsqlExecution info, not some possibly inaccessible rollup with the ComputeListener.

And I don't see how we can remove `runningOnRemoteCluster()` since we need to know in the refs Listener whether to populate the ComputeResponse with execution info metadata or not. The remote and local cluster refs listeners have to act differently. 

I also don't see how we can remove `isCCSListener` since the acquireCompute handler needs to know whether it is listener for a ComputeResponse with remote metadata or not, as again it needs to behave differently than a data-node handler listener. That's why I split them into separate methods in my earlier iteration of the PR, making the context or use case clear. 

Finally, another reason context is needed is because when the Status of a cluster gets set to SUCCESSFUL differs between remote and local clusters. For remote clusters, we mark it done when the remote ComputeResponse comes back. But for local, we have to wait until the coordinator is done (in case it has to run operators that aggregate/merge data from the remotes), so those get set in different places, and context/state is needed to know what code to execute in these methods that are shared for the 4 uses cases for which ComputeListener is used.

Again, I could be wrong and there's a cleaner way to do this, but I was not able to figure it out. I'd need some detailed guidance on how we solve the above problems and do local rollup accumulations like we do for profiles and warnings.

Reply:
Let's discuss this later.